src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1711,27 +1679,22 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
 
             var win32Resources = win32ResourcesStreamProvider?.GetOrCreateStream(diagnostics);
             var xmlDocumentationStream = xmlDocumentationStreamProvider?.GetOrCreateStream(diagnostics);
-            if (!this.Compile(
+            var success = this.Compile(
                 moduleBeingBuilt,
                 win32Resources,
                 xmlDocumentationStream,
                 emittingPdb: pdbStreamProvider != null,
                 diagnostics: diagnostics,
                 filterOpt: null,
-                cancellationToken: cancellationToken))
-            {
-                return ToEmitResultAndFree(diagnostics, success: false);
-            }
-
-            var hostDiagnostics = getHostDiagnostics?.Invoke() ?? ImmutableArray<Diagnostic>.Empty;
+                cancellationToken: cancellationToken);
 
-            diagnostics.AddRange(hostDiagnostics);
-            if (hostDiagnostics.Any(x => x.Severity == DiagnosticSeverity.Error))
+            if (((getHostDiagnostics != null) && !getHostDiagnostics(diagnostics, success)) ||
```

Comment:
This is remarkably hard to read.  Can we break this up into simpler (named) expressions so that we don' mix so many ands, ors and nots together?